sklearn/preprocessing/resample.py

```diff
@@ -0,0 +1,259 @@
+    Returns labels and indices where those labels appear in y.
+    >>> _collect_indices(np.array([1, 2, 2, 3, 3, 3, 4, 4, 4, 4]))
+    ... # doctest: +NORMALIZE_WHITESPACE, +ELLIPSIS
+    ((1, 2, 3, 4), ([0], [1, 2], [3, 4, 5], [6, 7, 8, 9]))
+    """
+    d = defaultdict(list)
+    for k, v in enumerate(y):
+        d[v].append(k)
+    labels, indices = zip(*sorted(d.items()))
+    return labels, indices
+
+
+def _circular_sample(initial_n_samples, target_n_samples, random_state=None):
+    """Sample without replacement from array in a loop.
+    Take each sample once per loop except for the last loop, which takes
+    (n % len(array)) samples at random. Results are returned unshuffled.
+    >>> _circular_sample(3, 8, random_state=46)
+    array([0, 0, 1, 1, 2, 2, 1, 0])
+    """
+    random_state = check_random_state(random_state)
+    n_full_sets = (int)(target_n_samples // initial_n_samples)
+    n_remainder = target_n_samples % initial_n_samples
+    sample_indices = np.empty(target_n_samples, dtype=int)
+    last_loop_index = n_full_sets * initial_n_samples
+    sample_indices[:last_loop_index] =\
+        np.repeat(np.arange(initial_n_samples), n_full_sets)
+    if n_remainder > 0:
+        sample_indices[last_loop_index:] = \
+            sample_without_replacement(initial_n_samples, n_remainder,
+                                       random_state=random_state)
+    return sample_indices
+
+
+def _fair_array_counts(n_samples, n_classes, random_state=None):
+    """Tries to fairly partition n_samples between n_classes.
+    If this cannot be done fairly, +1 is added `remainder` times
+    to the counts for random arrays until a total of `n_samples` is
+    reached.
+    >>> _fair_array_counts(5, 3, random_state=43)
+    array([2, 1, 2])
+    """
+    if n_classes > n_samples:
+        raise ValueError("The number of classes is greater"
+                         " than the number of samples requested")
+    sample_size = n_samples // n_classes
+    sample_size_rem = n_samples % n_classes
+    counts = np.repeat(sample_size, n_classes)
+    if sample_size_rem > 0:
+        counts[:sample_size_rem] += 1
+        # Shuffle so the class inbalance varies between runs
+        random_state = check_random_state(random_state)
+        random_state.shuffle(counts)
+    return counts
+
+
+def _scale_n_samples(scaling, n):
+    """Helper function to scale the number of samples."""
+    if scaling is None:
+        return n
+    else:
+        if isinstance(scaling, numbers.Number) and scaling < 0:
+            raise ValueError("Scaling must be nonnegative: %s" % scaling)
+        elif isinstance(scaling, float):
+            return scaling * n
+        elif isinstance(scaling, (numbers.Integral, np.integer)):
+            return scaling
+        else:
+            raise ValueError("Invalid value for scaling, must be "
+                             "float, int, or None: %s" % scaling)
+
+
+def weighted_sample(probas, n_samples, random_state=None):
```

Comment:
I suspect this should be preceded by `_` and non-public.

If it is to be public, it requires a `Returns` description